File: structural/C++/Proxy.cpp
#include <iostream>
#include <string>

class Image
{
protected:
	std::string name;

public:
	virtual void display() = 0;
};

class RealImage : public Image
{
public:
	RealImage(std::string imageName)
	{
		this->name = imageName;
		this->loadFromDisk(this->name);
	}

	void display()
	{
		std::cout << "Displaying " << this->name << " from memory.\n" << std::endl;
	}

private:
	void loadFromDisk(std::string imageName)
	{
		std::cout << "Loading image: " << imageName << " from disk." << std::endl;
	}
};

class ProxyImage : public Image
{
private:
	RealImage* realImage;

public:
	ProxyImage(std::string imageName): realImage(nullptr)
	{
		this->name = imageName;
	}

	void display()
	{
		if (this->realImage == nullptr)
		{
			this->realImage = new RealImage(this->name);
		}
		this->realImage->display();
	}
};

int main()
{
	Image* image = new ProxyImage("test_10mb.jpg");

	// image will be loaded from disk.
	image->display();

	// image will not be loaded from disk.
	image->display();

	system("PAUSE");

	return 0;
}

File: structural/PHP/Proxy.php
<?php

abstract class Image {

    protected $name;

    public abstract function display();

}

class RealImage extends Image {

    public function __construct($imageName) {
        $this->name = $imageName;
        $this->loadFromDisk($this->name);
    }

    // override
    public function display() {
        echo "Displaying " . $this->name . "from memory.\n\n";
    }

    public function loadFromDisk($imageName) {
        echo "Loading image: " . $imageName . " from disk.\n";
    }

}

class ProxyImage extends Image {

    private $realImage;

    public function __construct($imageName) {
        $this->name = $imageName;
    }

    // override
    public function display() {
        if ($this->realImage == null) {
            $this->realImage = new RealImage($this->name);
        }
        $this->realImage->display();
    }

}

echo ":: Proxy ::\n\n";

$image = new ProxyImage('test_10mb.jpg');

// image will be loaded from disk.
$image->display();

// image will not be loaded from disk.
$image->display();

?>